ledfx/virtuals.py

```diff
@@ -291,47 +315,46 @@ def set_preset(self, preset_info):
         self.set_effect(effect)
 
     def set_effect(self, effect):
-        self.lock.acquire()
-        if not self._devices:
-            error = f"Virtual {self.id}: Cannot activate, no configured device segments"
-            _LOGGER.warning(error)
-            raise ValueError(error)
+        with self.lock:
+            if not self._devices:
+                error = f"Virtual {self.id}: Cannot activate, no configured device segments"
+                _LOGGER.warning(error)
+                raise ValueError(error)
 
-        if (
-            self._config["transition_mode"] != "None"
-            and self._config["transition_time"] > 0
-        ):
-            self.transition_frame_total = (
-                self.refresh_rate * self._config["transition_time"]
-            )
-            self.transition_frame_counter = 0
-            self.clear_transition_effect()
+            if (
+                self._config["transition_mode"] != "None"
+                and self._config["transition_time"] > 0
+            ):
+                self.transition_frame_total = (
+                    self.refresh_rate * self._config["transition_time"]
+                )
+                self.transition_frame_counter = 0
+                self.clear_transition_effect()
 
-            if self._active_effect is None:
-                self._transition_effect = DummyEffect(self.pixel_count)
+                if self._active_effect is None:
+                    self._transition_effect = DummyEffect(self.pixel_count)
+                else:
+                    self._transition_effect = self._active_effect
             else:
-                self._transition_effect = self._active_effect
-        else:
-            # no transition effect to clean up, so clear the active effect now!
-            self.clear_active_effect()
-            self.clear_transition_effect()
+                # no transition effect to clean up, so clear the active effect now!
+                self.clear_active_effect()
+                self.clear_transition_effect()
 
-        self._active_effect = effect
-        self._active_effect.activate(self)
-        self._ledfx.events.fire_event(
-            EffectSetEvent(
-                self._active_effect.name,
-                self._active_effect.id,
-                self.active_effect.config,
-                self.id,
+            self._active_effect = effect
+            self._active_effect.activate(self)
+            self._ledfx.events.fire_event(
+                EffectSetEvent(
+                    self._active_effect.name,
+                    self._active_effect.id,
+                    self.active_effect.config,
+                    self.id,
+                )
             )
-        )
-        self.lock.release()
-        try:
-            self.active = True
-        except RuntimeError:
-            self.active = False
-            raise
+            try:
```

Comment:
I think this is a problem, the try / except clause around self.active = True has been moved inside the with self.lock context.

acitve is a setter function, and it internally under certain conditions can call self.activate or self.deactivate

There is a lot of code branching and calling below there, and I cannot be confident there are not paths that will hit a lock stansa. Can this code be move back outside of the lock context?